internal/registry/registry.go

```diff
@@ -402,37 +402,36 @@ func (r *Registry) syncStartDiscover(peerID peer.ID, discoAddr string, errCh cha
 	go func() {
 		discoData, discoErr := r.discover(peerID, discoAddr)
 		r.actions <- func() {
-			r.syncEndDiscover(discoAddr, discoData, discoErr, errCh)
-			r.discoWait.Done()
+			defer close(errCh)
+			defer r.discoWait.Done()
+
+			// Update discovery completion time.
+			r.discoTimes[discoAddr] = time.Now()
+			if discoErr != nil {
+				errCh <- discoErr
+				return
+			}
+
+			info := &ProviderInfo{
+				AddrInfo:      discoData.AddrInfo,
+				DiscoveryAddr: discoAddr,
+			}
+			if err := r.syncRegister(context.Background(), info); err != nil {
```

Comment:
might be good also for this to have a timeout?

Reply:
This should use the configured discovery timeout.  Fixed.